cloud-controller-manager/do/firewall_controller.go

```diff
@@ -0,0 +1,365 @@
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
+)
+
+const (
+	// Interval of synchronizing service status from apiserver.
+	serviceSyncPeriod = 30 * time.Second
+	// Frequency at which the firewall controller runs.
+	firewallReconcileFrequency = 5 * time.Minute
+)
+
+var (
+	allowAllOutboundRules = []godo.OutboundRule{
+		{
+			Protocol:  "tcp",
+			PortRange: "all",
+			Destinations: &godo.Destinations{
+				Addresses: []string{"0.0.0.0/0", "::/0"},
+			},
+		},
+		{
+			Protocol:  "udp",
+			PortRange: "all",
+			Destinations: &godo.Destinations{
+				Addresses: []string{"0.0.0.0/0", "::/0"},
+			},
+		},
+		{
+			Protocol: "icmp",
+			Destinations: &godo.Destinations{
+				Addresses: []string{"0.0.0.0/0", "::/0"},
+			},
+		},
+	}
+)
+
+// firewallCache stores a cached firewall and mutex to handle concurrent access.
+type firewallCache struct {
+	mu       *sync.RWMutex // protects firewall.
+	firewall *godo.Firewall
+}
+
+// firewallManagerOp manages the interaction with the DO Firewalls API.
+type firewallManagerOp struct {
+	client             *godo.Client
+	fwCache            firewallCache
+	workerFirewallName string
+	workerFirewallTags []string
+}
+
+// FirewallManager retrieves and stores firewall representations.
+type FirewallManager interface {
+	// Get returns the current public access firewall representation (i.e., the DO Firewall object)
+	// if it exists and nil if it does not exist.
+	Get(ctx context.Context) (*godo.Firewall, error)
+
+	// Set applies the given inbound rules to the public access firewall.
+	Set(ctx context.Context, firewallRequest *godo.FirewallRequest) error
+}
+
+// FirewallController helps to keep cloud provider service firewalls in sync.
+type FirewallController struct {
+	kubeClient         clientset.Interface
+	client             *godo.Client
+	workerFirewallTags []string
+	workerFirewallName string
+	serviceLister      corelisters.ServiceLister
+	fwManager          FirewallManager
+}
+
+// NewFirewallController returns a new firewall controller to reconcile public access firewall state.
+func NewFirewallController(
+	ctx context.Context,
+	kubeClient clientset.Interface,
+	client *godo.Client,
+	serviceInformer coreinformers.ServiceInformer,
+	fwManager FirewallManager,
+	workerFirewallTags []string,
+	workerFirewallName string,
+) *FirewallController {
+	fc := &FirewallController{
+		kubeClient:         kubeClient,
+		client:             client,
+		workerFirewallTags: workerFirewallTags,
+		workerFirewallName: workerFirewallName,
+		fwManager:          fwManager,
+	}
+
+	serviceInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(cur interface{}) {
+				ctx, cancel := context.WithTimeout(context.Background(), 90*time.Second)
+				defer cancel()
+				err := fc.ensureReconciledFirewall(ctx)
+				if err != nil {
+					klog.Errorf("failed to reconcile worker firewall: %s", err)
```

Comment:
Can we distinguish the three handler cases by something in the log message to tell us if a failure is about an addition, update, or deletion?